### PR TITLE
nodejs/master merge

### PR DESCRIPTION
**Pull request by kunalspathak**
Merge 642baf4 as of 2017-03-28

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines][]

##### Affected core subsystem(s)

src, test, chakrashim

[commit guidelines]: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines

### CODE REVIEW

**Comment by kunalspathak** [comment]
@digitalinfinity  - Please review.

Linux CI - https://ci.nodejs.org/job/chakracore-test-linux/3/console
Windows CI - https://ci.nodejs.org/job/chakracore-test-windows/48/console
Mac CI - https://ci.nodejs.org/job/chakracore-test-osx/2/console

File: deps/chakrashim/src/v8functiontemplate.cc [hunk 1]
```diff
@@ -264,12 +264,14 @@ class FunctionTemplateData : public TemplateData {
...
-
-        if (JsSetProperty(*prototype,
-                          iso->GetCachedPropertyIdRef(
-                            jsrt::CachedPropertyIdRef::constructor),
-                          function, false) != JsNoError ||
-            JsSetProperty(function,
+        if (jsrt::DefineProperty(*prototype,
+            iso->GetCachedPropertyIdRef(
+                jsrt::CachedPropertyIdRef::constructor),
+            jsrt::PropertyDescriptorOptionValues::True,
```
Comment (digitalinfinity) [inline deps/chakrashim/src/v8functiontemplate.cc]: Nit: can you comment the individual bool params so its easier to read which properties are being set on the property descriptor without having to refer to the definition of jsrt::DefineProperty?

**Review by digitalinfinity — APPROVED** [review]
Reviewed 7c83517 and it looks fine. The rest looks like a vanilla merge so LGTM

**Comment by digitalinfinity** [comment]
Btw @kunalspathak the Linux CI is not working yet. @joaocgreis had some suggestions, and I'm discussing a potential change to ChakraCore with @obastemur which could help resolve this issue. I'm not sure what's up with the macOS CI though

**Comment by joaocgreis** [comment]
Linux CI - As @digitalinfinity mentioned, not working yet. Just ignore it for now.
Windows CI - All green where it should be (I see https://github.com/Microsoft/ChakraCore/pull/2743 was merged upstream, so it'll soon be completely green).
OSX CI - [3 legitimate failures.](https://ci.nodejs.org/job/chakracore-test-osx/2/nodes=osx1010/console) These 3 failures are the [same as in `xplat`](https://ci.nodejs.org/job/chakracore-test-osx/1/nodes=osx1010/console), so this PR does not introduce any new failure. We should investigate to mark those tests to be skipped or fix them.

**Comment by kunalspathak** [comment]
Also took latest chakracore release/2.0. 

CI - https://ci.nodejs.org/job/chakracore-test-windows/49/

**Comment by jianchun** [comment]
@kunalspathak Could you consider sync one more commit from ChakraCore release/2.0, Lei's PR 2763, pushed this morning? Fixes a lifetime issue that causes random crash.

**Comment by kunalspathak** [comment]
@jianchun  - Done.

CI - https://ci.nodejs.org/job/chakracore-test-windows/50/

**Comment by jianchun** [comment]
@kunalspathak Thanks!

**Comment by digitalinfinity** [comment]
@joaocgreis looks like Microsoft/ChakraCore#2743 and Microsoft/ChakraCore#2764 have landed in Node-ChakraCore with this merge. Kunal's last Windows CI run shows that VC++ Build Tools CI is passing. We should be able to get the Linux CI to also work with the current version

**Pull request merged**